tests/test_trial.py

```diff
@@ -764,9 +765,328 @@ def test_study_id():
     assert trial._study_id == trial.study._study_id
 
 
-def test_frozen_trial_validate():
+def test_frozen_trial_eq_ne():
+    # type: () -> None
+
+    trial = _create_frozen_trial()
+
+    trial_other = copy.copy(trial)
+    assert trial == trial_other
+
+    trial_other.value = 0.3
+    assert trial != trial_other
+
+
+def test_frozen_trial_lt():
+    # type: () -> None
+
+    trial = _create_frozen_trial()
+
+    trial_other = copy.copy(trial)
+    assert not trial < trial_other
+
+    trial_other.number = trial.number + 1
+    assert trial < trial_other
+    assert not trial_other < trial
+
+    with pytest.raises(TypeError):
+        trial < 1
+
+    assert trial <= trial_other
+    assert not trial_other <= trial
+
+    with pytest.raises(TypeError):
+        trial <= 1
+
+    # A list of FrozenTrials is sortable.
+    trials = [trial_other, trial]
+    trials.sort()
+    assert trials[0] is trial
+    assert trials[1] is trial_other
+
+
+def _create_frozen_trial():
+    # type: () -> FrozenTrial
+
+    return FrozenTrial(
+        number=0,
+        trial_id=0,
+        state=TrialState.COMPLETE,
+        value=0.2,
+        datetime_start=datetime.datetime.now(),
+        datetime_complete=datetime.datetime.now(),
+        params={"x": 10},
+        distributions={"x": UniformDistribution(5, 12)},
+        user_attrs={},
+        system_attrs={},
+        intermediate_values={},
+    )
+
+
+def test_frozen_trial_repr():
     # type: () -> None
 
+    trial = FrozenTrial(
+        number=0,
+        trial_id=0,
+        state=TrialState.COMPLETE,
+        value=0.2,
+        datetime_start=datetime.datetime.now(),
+        datetime_complete=datetime.datetime.now(),
+        params={"x": 10},
+        distributions={"x": UniformDistribution(5, 12)},
+        user_attrs={},
+        system_attrs={},
+        intermediate_values={},
+    )
+
+    assert trial == eval(repr(trial))
+
+
+@parametrize_storage
+def test_frozen_trial_sampling(storage_init_func):
+    # type: (Callable[[], storages.BaseStorage]) -> None
+
+    def objective(trial: BaseTrial) -> float:
+
+        a = trial.suggest_uniform("a", 0, 10)
+        b = trial.suggest_loguniform("b", 0.1, 10)
+        c = trial.suggest_discrete_uniform("c", 0, 10, 1)
+        d = trial.suggest_int("d", 0, 10)
+        e = trial.suggest_categorical("e", [0, 1, 2])
+        f = trial.suggest_int("f", 1, 10, log=True)
+
+        return a + b + c + d + e + f
```

Comment:
mypy`s error shows

```python
tests/test_trial.py:859: error: Unsupported operand types for + ("float" and "None")
tests/test_trial.py:859: error: Unsupported operand types for + ("float" and "str")
tests/test_trial.py:859: note: Right operand is of type "Union[None, bool, int, float, str]"
Found 2 errors in 1 file (checked 183 source files)
```

Reply:
I cannot understand why this error happens...

Reply:
This is because the type of returned value for `suggest_categorical` is `CategoricalChoiceType = Union[None, bool, int, float, str]`. How about asserting the `e` is `float` in Line 858 as follows.

```python
assert isinstance(e, float)
```

Reply:
Sorry for my poor understanding of `mypy`. (I'm still confused this technique as your suggestion, would you mind letting me know some reference about it?)


The returned value is not `float`, so `assert isinstance(e, int)` works. Anyway, mypy's error does not appear.
@HideakiImamura  Many thanks for your quick response!

Reply:
Okay, probably [this part](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-attribute-exists-in-each-union-item-union-attr)...

> You can often work around these errors by using assert isinstance(obj, ClassName) or assert obj is not None to tell mypy that you know that the type is more specific than what mypy thinks.

Reply:
Thanks for the reply and sorry for my late response.

> Okay, probably this part...

Yes, you are right! We have to insert some assertions like above to pass `mypy` sometimes. :)